src/systems/tips.js

```diff
@@ -0,0 +1,274 @@
+    if (userinput.activeSets.has(sets.cursorHoldingInteractable)) return INVALID;
+    return scene.is("copresent") ? FINISH : VALID;
+  },
+  object_grab: function(userinput, scene, mediaCounter) {
+    if (scene.is("frozen")) return INVALID;
+    if (mediaCounter.count() === 0) return INVALID;
+    if (userinput.activeSets.has(sets.cursorHoldingPen)) return INVALID;
+    if (userinput.activeSets.has(sets.cursorHoldingCamera)) return INVALID;
+    if (userinput.activeSets.has(sets.cursorHoldingInteractable)) return FINISH;
+    return VALID;
+  },
+  object_pin: function(userinput, scene, mediaCounter, store) {
+    if (!scene.is("frozen")) return INVALID;
+    if (mediaCounter.count() === 0) return INVALID;
+    if (store && store.state.activity.hasPinned) return FINISH;
+    return VALID;
+  },
+  object_zoom: function(userinput) {
+    if (userinput.activeSets.has(sets.cursorHoldingPen)) return INVALID;
+    if (userinput.activeSets.has(sets.cursorHoldingCamera)) return INVALID;
+    if (!userinput.activeSets.has(sets.cursorHoldingInteractable)) return INVALID;
+    if (userinput.get(paths.actions.cursor.modDelta)) return FINISH;
+    return VALID;
+  },
+  object_scale: function(userinput) {
+    if (userinput.activeSets.has(sets.cursorHoldingPen)) return INVALID;
+    if (userinput.activeSets.has(sets.cursorHoldingCamera)) return INVALID;
+    if (!userinput.activeSets.has(sets.cursorHoldingInteractable)) return INVALID;
+    if (userinput.get(paths.actions.cursor.scaleGrabbedGrabbable)) return FINISH;
+    return VALID;
+  },
+  pen_color: function(userinput) {
+    if (!userinput.activeSets.has(sets.cursorHoldingPen)) return INVALID;
+    if (userinput.get(paths.actions.cursor.penNextColor) || userinput.get(paths.actions.cursor.penPrevColor)) {
+      return FINISH;
+    }
+    return VALID;
+  },
+  pen_size: function(userinput) {
+    if (!userinput.activeSets.has(sets.cursorHoldingPen)) return INVALID;
+    if (userinput.get(paths.actions.cursor.scalePenTip)) return FINISH;
+    return VALID;
+  },
+  pen_mode: function(userinput) {
+    if (!userinput.activeSets.has(sets.cursorHoldingPen)) return INVALID;
+    return VALID;
+  },
+  freeze_mode: function(userinput, scene) {
+    return scene.is("frozen") ? VALID : INVALID;
+  },
+  video_share_mode: function(userinput, scene) {
+    return scene.is("sharing_video") ? VALID : INVALID;
+  },
+  mirror_mode: function(userinput, scene) {
+    return scene.is("mirroring") ? VALID : INVALID;
+  },
+  mute_mode: function(userinput, scene) {
+    return scene.is("muted") ? VALID : INVALID;
+  }
+};
+
+AFRAME.registerSystem("tips", {
+  init: function() {
+    this.activeTips = {};
+    this._finishedScopes = {};
+    this._performStep = this._performStep.bind(this);
+
+    if (localStorage.getItem(LOCAL_STORAGE_KEY) === null) {
+      localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify({}));
+    }
+
+    this._tickCount = 0;
+  },
+
+  tick: function() {
```

Comment:
Seems a shame we are going to be running this tick function in the very common case that you are continuing to use hubs and have already completed all the tips. Made some suggestions above about avoiding the work if we detect that we have no tips to display this session.

Reply:
the top tips are always "in-play", since we never FINISH them. (arguably we could explore that later, I was not sure if it made sense or not since the top tips basically give you a clear way to escape out of "modes" if you never learned or forgot how to.) the bottom tips, which are more sequences and one-time, have an early-exit optimization where it just checks `finishedScopes` at least so it doesn't have to walk + validate them all once you've completed them.